src/service/history-impl.js

```diff
@@ -525,6 +545,28 @@ export class HistoryBindingNatural_ {
   }
 
   /**
+   * If this is a hash update the choice of `location.replace` vs
+   * `history.replaceState` is important. Due to bugs, not every browser
+   * triggers `:target` pseudo-class when `replaceState` is called.
+   * See http://www.zachleat.com/web/moving-target/ for more details.
+   * location.replace will trigger a `popstate` event, we temporarily
+   * disable handling it.
+   * @param {string} target
+   *
+   * @override
+   */
+  replaceStateForTarget(target) {
+    this.whenReady_(() => {
+      let hash = target.indexOf('#') == 0 ? target : `#${target}`;
+      this.ignoreUpcomingPopstate_ = true;
+      // TODO(mkhatib, #6095): Chrome iOS will add extra states for location.replace.
+      this.win.location.replace(hash);
+      this.ignoreUpcomingPopstate_ = false;
+      this.historyReplaceState_();
```

Comment:
Not sure what do you mean navigation to the same URL. We're going to a new URL here, where do we get access to the `previous` Url?

The `popstate` happens for any call for `location.replace` in this case we don't want that to be considered a `popstate`. 

Mind elaborating? I've tried testing what you suggested but couldn't get it to work.